File: ServiGo/backend/apps/usuarios/models/reserva.py
from django.db import models

from ..data.estados_reserva import EstadoReserva
from .profesional import Profesional
from .servicio import Servicio
from .usuario import Usuario


class Reserva(models.Model):
    profesional = models.ForeignKey(
        Profesional,
        on_delete=models.CASCADE,
        related_name="reservas"
    )

    servicio = models.ForeignKey(
        Servicio,
        on_delete=models.CASCADE,
        related_name="reservas"
    )

    cliente = models.ForeignKey(
        Usuario,
        on_delete=models.CASCADE,
        related_name="reservas"
    )

    fecha = models.DateField()
    hora = models.TimeField()

    estado = models.CharField(
        max_length=20,
        choices=EstadoReserva.choices,
        default=EstadoReserva.PENDIENTE
    )

    notas = models.TextField(blank=True)

    creada_en = models.DateTimeField(
        auto_now_add=True
    )

    actualizada_en = models.DateTimeField(
        auto_now=True
    )

    def __str__(self):
        return (
            f"{self.cliente.username} - "
            f"{self.servicio.nombre} - "
            f"{self.fecha}"
        )

File: backend/src/models/Reserva.ts
import { Schema, model, Document, Types } from "mongoose";
import { ESTADOS_RESERVA, EstadoReserva } from "../data/estadosReserva";

export interface IReserva extends Document {
  profesional: Types.ObjectId;
  servicio: Types.ObjectId;
  cliente: Types.ObjectId;
  fecha: string;
  hora: string;
  estado: EstadoReserva;
  notas?: string;
  creadaEn: Date;
  actualizadaEn: Date;
}

const reservaSchema = new Schema<IReserva>(
  {
    profesional: { type: Schema.Types.ObjectId, ref: "Profesional", required: true },
    servicio: { type: Schema.Types.ObjectId, ref: "Servicio", required: true },
    cliente: { type: Schema.Types.ObjectId, ref: "Usuario", required: true },
    fecha: { type: String, required: true },
    hora: { type: String, required: true },
    estado: { type: String, enum: Object.values(ESTADOS_RESERVA), default: ESTADOS_RESERVA.PENDIENTE },
    notas: { type: String, default: "" },
  },
  { timestamps: { createdAt: "creadaEn", updatedAt: "actualizadaEn" } }
);

reservaSchema.index({ profesional: 1, fecha: 1, hora: 1 });

export const Reserva = model<IReserva>("Reserva", reservaSchema);
